File: javascript/problema1/main.js
import fs from "node:fs";

// Lectura del archivo
const textoSucio = fs.readFileSync("../texto_prueba.txt", "utf8");

// Se limpian los signos de puntuación para hacer el conteo de palabras
// mediante una expresión regular
const textoLimpio = textoSucio
  .replace(/[!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~]/g, "")
  .toLowerCase();

// Diccionario que almacenará la cantidad de veces que se repite cada palabra
const conteoPalabras = {};

// Separar el texto en palabras y contar cuántas veces aparece cada una
textoLimpio.split(" ").forEach((palabra) => {
  if (conteoPalabras[palabra]) {
    conteoPalabras[palabra] += 1;
  } else {
    conteoPalabras[palabra] = 1;
  }
});

// Ordenamos el diccionario por la cantidad de veces que aparece cada palabra
const conteoPalabrasSort = Object.entries(conteoPalabras).sort(
  ([, a], [, b]) => b - a
);

// Imprimir el resultado
for (let i = 0; i < 10; i++) {
  console.log(
    `${i + 1}: ${conteoPalabrasSort[i][0]} ${conteoPalabrasSort[i][1]} veces`
  );
}

File: ruby/problema1/main.rb
# Lectura del archivo
archivo = File.open('../../texto_prueba.txt', 'r')
texto_sucio = archivo.read
archivo.close

start_time = Time.now
# Se limpian los signos de puntuación para hacer el conteo de palabras
# mediante una expresión regular
texto_limpio = texto_sucio.gsub(/[[:punct:]]/, '').downcase

conteo_palabras = {}

# Separar el texto en palabras y contar cuántas veces aparece cada una
# slit por defecto separa por espacios
texto_limpio.split.each do |palabra|
    if conteo_palabras[palabra]
        conteo_palabras[palabra] += 1
    else
        conteo_palabras[palabra] = 1
    end
end

# Ordenamos el diccionario por la cantidad de veces que aparece cada palabra
conteo_palabras_sort = conteo_palabras.sort_by { |palabra, conteo| - conteo }

end_time = Time.now

execution_time = end_time - start_time
puts "Tiempo de ejecución: #{execution_time} segundos"

# Imprimir el resultado
# Los primeros 10
# Para imprimir con el indice se debe usar each_with_index
conteo_palabras_sort.first(10).each_with_index do |(palabra, cantidad), index|
    puts "#{index + 1}: #{palabra} #{cantidad} veces"
end